controller/apps_kubeclient.go

```diff
@@ -0,0 +1,717 @@
+	metricsURL, err := getMetricsURLFromAPIURL(clusterURL)
+	if err != nil {
+		return nil, err
+	}
+	// Create MetricsClient for talking with Hawkular API
+	metrics, err := newMetricsClient(metricsURL, kubeToken)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeClient := new(KubeClient)
+	kubeClient.config = &config
+	kubeClient.clientset = clientset
+	kubeClient.userNamespace = userNamespace
+	kubeClient.metrics = metrics
+
+	// Get environments from config map
+	envMap, err := kubeClient.getEnvironmentsFromConfigMap()
+	if err != nil {
+		return nil, err
+	}
+	kubeClient.envMap = envMap
+	return kubeClient, nil
+}
+
+// GetSpace returns a space matching the provided name, containing all applications that belong to it
+func (kc *KubeClient) GetSpace(spaceName string) (*app.SimpleSpace, error) {
+	// Get BuildConfigs within the user namespace that have a matching 'space' label
+	// This is similar to how pipelines are displayed in fabric8-ui
+	// https://github.com/fabric8-ui/fabric8-ui/blob/master/src/app/space/create/pipelines/pipelines.component.ts
+	buildconfigs, err := kc.getBuildConfigs(spaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all applications in this space using BuildConfig names
+	var apps []*app.SimpleApp
+	for _, bc := range buildconfigs {
+		appn, err := kc.GetApplication(spaceName, bc)
+		if err != nil {
+			return nil, err
+		}
+		apps = append(apps, appn)
+	}
+
+	result := &app.SimpleSpace{
+		Applications: apps, // TODO UUID
+	}
+
+	return result, nil
+}
+
+// GetApplication retrieves an application with the given space and application names, with the status
+// of that application's deployment in each environment
+func (kc *KubeClient) GetApplication(spaceName string, appName string) (*app.SimpleApp, error) {
+	// Get all deployments of this app for each environment in this space
+	var deployments []*app.SimpleDeployment
+	for envName := range kc.envMap {
+		deployment, err := kc.GetDeployment(spaceName, appName, envName)
+		if err != nil {
+			return nil, err
+		} else if deployment != nil {
+			deployments = append(deployments, deployment)
+		}
+	}
+
+	result := &app.SimpleApp{
+		Name:     &appName, // TODO UUID
+		Pipeline: deployments,
+	}
+	return result, nil
+}
+
+// ScaleDeployment - scale a deployment
+func (kc *KubeClient) ScaleDeployment(spaceName string, appName string, envName string, deployNumber int) (int, error) {
+	// Look up DeploymentConfig corresponding to the application name in the provided environment
+	dc, err := kc.getDeploymentConfig(envName, appName, spaceName)
+	if err != nil {
+		return -1, err
+	} else if len(dc) == 0 {
+		return -1, nil
+	}
+
+	spec, ok := dc["spec"].(map[interface{}]interface{})
+	if !ok {
+		return -1, errors.New("Invalid deployment config returned from endpoint: missing 'spec'")
+	}
+
+	oldReplicas, ok := spec["replicas"].(int)
+	if !ok {
+		return -1, errors.New("Invalid deployment config returned from endpoint: missing 'replicas'")
```

Comment:
if the method returns `(*int, error)`, then you can do `return nil, errors.New("...")` here

Reply:
done